### PR TITLE
Add instance as data source

### PR DESCRIPTION
**Pull request by bhuvaneswaribalaji07**
Changes to Add instance as data-source.
Improved error handling for environment, group and topic datasources.

### CODE REVIEW

**Review by kasparAxual — CHANGES_REQUESTED** [review]

File: templates/data-sources/instance.md.tmpl [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+```hcl
+data "axual_instance" "testInstance" {
+ name = "testInstance"
+}
+```
+
+## Argument Reference
+
+- name - (Required) The group name.
```
Comment (kasparAxual) [inline templates/data-sources/instance.md.tmpl:17]: instance?

File: templates/data-sources/instance.md.tmpl [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+## Attribute Reference
+
+This data source exports the following attributes in addition to the one listed above:
+
+- id group unique identifier.
+- short_name The instance short name.
+- description The group description.
```
Comment (kasparAxual) [inline templates/data-sources/instance.md.tmpl:25]: instance?

File: templates/data-sources/instance.md.tmpl [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+## Argument Reference
+
+- name - (Required) The group name.
+
+## Attribute Reference
+
+This data source exports the following attributes in addition to the one listed above:
+
+- id group unique identifier.
```
Comment (kasparAxual) [inline templates/data-sources/instance.md.tmpl:23]: instance?

File: templates/guides/data-sources.md [hunk 4]
```diff
@@ -131,3 +132,24 @@ resource "axual_application_access_grant_approval" "logs_producer_produce_to_log
   application_access_grant = "data.logs_producer_produce_to_logs_in_dev.id"
 }
 ```
+
+- To define a `axual_group` data source, provide the group name:
```
Comment (kasparAxual) [inline templates/guides/data-sources.md:136]: instance?

File: docs/data-sources/instance.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+```hcl
+data "axual_instance" "testInstance" {
+ name = "testInstance"
+}
+```
+
+## Argument Reference
+
+- name - (Required) The group name.
```
Comment (kasparAxual) [inline docs/data-sources/instance.md:17]: instance?

File: docs/data-sources/instance.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+## Attribute Reference
+
+This data source exports the following attributes in addition to the one listed above:
+
+- id group unique identifier.
+- short_name The instance short name.
+- description The group description.
```
Comment (kasparAxual) [inline docs/data-sources/instance.md:25]: instance?

File: internal/provider/data_source_instance.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				MarkdownDescription: "Instance's unique identifier",
+				Computed:            true,
+			},
+			"name": schema.StringAttribute{
+				MarkdownDescription: "Instance's name",
+				Required:            true,
+				Validators: []validator.String{
+					stringvalidator.LengthBetween(3, 80),
```
Comment (kasparAxual) [inline internal/provider/data_source_instance.go:53]: Why 80? In Api max is 50.

File: internal/provider/data_source_instance.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+					stringvalidator.RegexMatches(regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9 ._-]*$`), "can only contain letters, numbers, dots, spaces, dashes and underscores, but cannot begin with an underscore, dot, space or dash"),
```
Comment (kasparAxual) [inline internal/provider/data_source_instance.go:54]: The API Java regex allows the string to start with a separator (dot, dash, underscore, or space).
The Terraform Go regex requires the string to start with an alphanumeric character (A-Za-z0-9).
Please change Terraform regex to match API Java regex.

File: internal/provider/data_source_instance.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read instance by name, got error: %s", err))
+		return
+	}
+
+	instance, err2 := d.provider.client.GetInstance(instanceByName.Uid)
+	if err2 != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read instance, got error: %s", err2))
+		return
+	}
+
```
Comment (kasparAxual) [inline internal/provider/data_source_instance.go:90]: Please add a test what happens if the instance is not found

File: internal/provider/data_source_group.go [hunk 10]
```diff
@@ -88,6 +88,15 @@ func (d *groupDataSource) Read(ctx context.Context, req datasource.ReadRequest,
 		return
 	}
 
+	// Check if Embedded or group is nil or empty
+	if len(groupByName.Embedded.Groups) == 0 {
```
Comment (kasparAxual) [inline internal/provider/data_source_group.go:92]: Please add a test what happens if the group is not found

File: internal/provider/data_source_topic.go [hunk 11]
```diff
@@ -106,6 +106,15 @@ func (d *topicDataSource) Read(ctx context.Context, req datasource.ReadRequest,
 		return
 	}
 
+	// Check if Embedded or topic is nil or empty
+	if len(topicByName.Embedded.Topics) == 0 {
```
Comment (kasparAxual) [inline internal/provider/data_source_topic.go:110]: Please add a test what happens if the topic is not found

File: internal/provider/data_source_environment.go [hunk 12]
```diff
@@ -127,6 +127,15 @@ func (d *environmentDataSource) Read(ctx context.Context, req datasource.ReadReq
 		return
 	}
 
+	// Check if Embedded or environment is nil or empty
+	if len(environmentByName.Embedded.Environments) == 0 {
```
Comment (kasparAxual) [inline internal/provider/data_source_environment.go:131]: Please add a test what happens if the env is not found

**Review by kasparAxual — CHANGES_REQUESTED** [review]

File: internal/provider/data_source_instance.go [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+	instanceByName, err := d.provider.client.GetInstanceByName(data.Name.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read instance by name, got error: %s", err))
+		return
+	}
+
+	instance, err2 := d.provider.client.GetInstance(instanceByName.Uid)
+	if err2 != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read instance, got error: %s", err2))
```
Comment (tunji-axual) [inline internal/provider/data_source_instance.go:88]:
```suggestion
		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read instance with ID '%s', got error: %s", instanceByName.Uid, err2))
```

File: internal/tests/InstanceDataSource/instance_data_source_test.go [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+					resource.TestCheckResourceAttr("data.axual_instance.dta", "name", "Dev Test Acceptance"),
+					resource.TestCheckResourceAttr("data.axual_instance.dta", "description", "Dev Test Acceptance"),
+					resource.TestCheckResourceAttr("data.axual_instance.dta", "short_name", "dta"),
+					resource.TestCheckResourceAttr("data.axual_instance.dta", "id", "4b0f204ede6542dfae6bf836f8185c5e"),
+				),
+			},
+			{
+				Config:      GetProvider() + GetFile("axual_instance_not_found.tf"),
+				ExpectError: regexp.MustCompile("Resource Not Found: No Instance resources found with name 'non_existent_resource'"),
+			},
```
Comment (tunji-axual) [inline internal/tests/InstanceDataSource/instance_data_source_test.go:29]: Can we add a test case for invalid name 

```suggestion
			},
			{
				// Invalid name attribute
				Config:      GetProvider() + GetFile("axual_instance_invalid_name.tf"),
				ExpectError: regexp.MustCompile("Invalid attribute value: The name must be between 3 and 50 characters."),
			},
```

**Review by tunji-axual — APPROVED** [review]

**Review by danny-akp — APPROVED** [review]

**Pull request merged**